incorporating-accessibility-into-your-dev-process/focus-tests/solution/src/App.tsx

```diff
@@ -1,58 +1,117 @@
+  isValid: boolean;
+  lastName: string | undefined;
+}
 
 const App = () => {
+  const firstNameErrorId = useId();
   const firstNameRef = useRef<HTMLInputElement>(null);
   const [firstNameValue, setFirstNameValue] = useState("");
 
+  const [submittedFormState, setSubmittedFormState] = useState<FormState>();
+
+  const lastNameErrorId = useId();
   const lastNameRef = useRef<HTMLInputElement>(null);
   const [lastNameValue, setLastNameValue] = useState("");
 
-  const previousFullName = useRef<string | null>();
-
-  const fullName = firstNameValue && lastNameValue ? `${firstNameValue} ${lastNameValue}` : "";
-
   function handleSubmit(event: FormEvent) {
     event.preventDefault();
 
+    // Keep track of when the form has been submitted
+    setSubmittedFormState({
+      firstName: firstNameRef.current?.value,
+      isValid: (event.target as HTMLFormElement).checkValidity(),
+      lastName: firstNameRef.current?.value,
+    });
+
     // Give focus to the first input element that isn’t valid
-    if (firstNameRef.current && firstNameRef.current.checkValidity() === false) {
+    if (
+      firstNameRef.current &&
+      firstNameRef.current.checkValidity() === false
+    ) {
       firstNameRef.current.focus();
-    } else if (lastNameRef.current && lastNameRef.current.checkValidity() === false) {
+    } else if (
+      lastNameRef.current &&
+      lastNameRef.current.checkValidity() === false
+    ) {
       lastNameRef.current.focus();
     }
+  }
 
-    // Let the user know they’ve updated the full name
-    if (previousFullName.current !== fullName) {
-      if (fullName) {
-        alert(`Update full name from “${previousFullName.current || ""}” to “${fullName}”`);
-      }
-      previousFullName.current = fullName;
+  function setSuccessRef(element: HTMLParagraphElement) {
+    if (element) {
+      element.focus();
     }
   }
 
   return (
-    <form noValidate onSubmit={handleSubmit}>
-      <p>
-        <label>
-          First name:
-          <input id="firstName" onChange={event => { setFirstNameValue(event.target.value) }} ref={firstNameRef} required />
-        </label>
-      </p>
-      <p>
-        <label>
-          Last name:
-          <input id="lastName" onChange={event => { setLastNameValue(event.target.value) }} ref={lastNameRef} required />
-        </label>
-      </p>
-      <p>
-        <button type="submit">Validate full name</button>
-      </p>
-      <hr />
-      <p>
-        {"Full name: "}
-        <output htmlFor="firstName lastName">{fullName || "incomplete name"}</output>
-      </p>
-    </form>
+    <>
+      {submittedFormState?.isValid ? (
+        <p ref={setSuccessRef} tabIndex={0}>
+          Success! You’ve submitted the form with this full name:{" "}
+          {`${firstNameValue} ${lastNameValue}`}
+        </p>
+      ) : (
+        <form noValidate onSubmit={handleSubmit}>
+          <p>
+            <label>
+              First name:
+              <input
```

Comment:
I think these inputs need an `aria-errormessage` on them

`aria-errormessage={ hasError ? firstNameErrorId : undefined}`

(pretend `hasError` exists)

Reply:
Maybe I’ll cover this in my training. `aria-errormessage ` [doesn’t have great support](https://a11ysupport.io/tech/aria/aria-errormessage_attribute) and it doesn’t accomplish anything more than `aria-invalid` and `aria-describedby` combined.